mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs

```diff
@@ -1897,5 +1897,72 @@ internal static void SetLastWin32Error (int error)
 			throw new NotSupportedException();
 		}
 #endif
+
+		internal class MarshalerInstanceKeyComparer : IEqualityComparer<(Type, string)> {
+			public bool Equals ((Type, string) lhs, (Type, string) rhs) {
+				return lhs.CompareTo(rhs) == 0;
+			}
+
+			public int GetHashCode ((Type, string) key) {
+				return key.GetHashCode ();
+			}
+		}
+
+		internal static Dictionary<(Type, string), ICustomMarshaler> MarshalerInstanceCache = null;
+		internal static object MarshalerInstanceCacheLock = new object ();
+
+		internal static ICustomMarshaler GetCustomMarshalerInstance (Type type, string cookie) {
+			var key = (type, cookie);
+
+			LazyInitializer.EnsureInitialized (
+				ref MarshalerInstanceCache, 
+				() => new Dictionary<(Type, string), ICustomMarshaler> (new MarshalerInstanceKeyComparer ())
+			);
+
+			ICustomMarshaler result;
+			bool gotExistingInstance;
+			lock (MarshalerInstanceCacheLock)
+				gotExistingInstance = MarshalerInstanceCache.TryGetValue (key, out result);
+
+			if (!gotExistingInstance) {
+				MonoMethod getInstanceMethod;
+				try {
+					getInstanceMethod = (MonoMethod)type.GetMethod (
+						"GetInstance", BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.InvokeMethod,
+						null, new Type[] { typeof(string) }, null
+					);
+				} catch (AmbiguousMatchException) {
+					throw new ApplicationException ($"Custom marshaler '{type.FullName}' implements multiple static GetInstance methods that take a single string parameter.");
+				}
+
+				if ((getInstanceMethod == null) || 
+					(getInstanceMethod.ReturnType != typeof (ICustomMarshaler))) {
+					throw new ApplicationException ($"Custom marshaler '{type.FullName}' does not implement a static GetInstance method that takes a single string parameter and returns an ICustomMarshaler.");
+				}
+
+				Exception exc;
+				try {
+					result = (ICustomMarshaler)getInstanceMethod.InternalInvoke (null, new object[] { cookie }, out exc);
+				} catch (Exception e) {
+					// FIXME: mscorlib's legacyUnhandledExceptionPolicy is apparently 1, 
+					//  so exceptions are thrown instead of being passed through the outparam
+					exc = e;
```

Comment:
Just doing `throw e;` would be same as capture + throw

Reply:
doesn't that trash the stack?

Reply:
yes, `throw e;` resets the stack trace whereas `throw;` doesn't

Reply:
I wanted to try to preserve and propagate the original stack, because that is what .net framework does. I noticed .Throw does not do that, but it at least includes the original stack (below the current stack), which is better than nothing.

Reply:
ok, let's move the Throw () into the catch section

Reply:
If the icall is changed to actually write the exception into the var, the throw won't get used and then we'll blow away the stack. I guess I could put it in both places?

Reply:
ok, I missed that part